### PR TITLE
[Feature](partition) Support OLAP table null partition

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #xxx

1. for auto partition, support nullable partition column.
2. for auto list partition, support create real null partition for null values.
3. for auto range partition, null value will raise a error now. but maybe we can improve this in the future

doc pr: https://github.com/apache/doris-website/pull/452

### CODE REVIEW

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/vec/sink/vrow_distribution.cpp [hunk 1]
```diff
@@ -247,29 +256,34 @@ Status VRowDistribution::_generate_rows_distribution_for_auto_partition(
...
-            vectorized::ColumnPtr range_left_col =
-                    block->get_by_position(partition_cols_idx[i]).column;
-            if (const auto* nullable =
-                        check_and_get_column<vectorized::ColumnNullable>(*range_left_col)) {
-                range_left_col = nullable->get_nested_column_ptr();
-                return_type = assert_cast<const vectorized::DataTypeNullable*>(return_type.get())
-                                      ->get_nested_type();
+        std::vector<const NullMap*> col_null_maps;
+        col_strs.resize(part_col_num);
+        col_null_maps.reserve(part_col_num);
```
[the remark this answers is not in the document]
Comment (zclllyybb) [reply be/src/vec/sink/vrow_distribution.cpp:265]: it has no perf diff here

**Comment by zclllyybb** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/analysis/PartitionExprUtil.java [hunk 2]
```diff
@@ -124,15 +124,23 @@ public static Map<String, AddPartitionClause> getAddPartitionClauseFromPartition
...
+        for (List<TNullableStringLiteral> partitionValueList : partitionValues) {
             PartitionKeyDesc partitionKeyDesc = null;
             String partitionName = "p";
             ArrayList<String> curPartitionValues = new ArrayList<>();
-            for (TStringLiteral tStringLiteral : partitionValueList) {
-                curPartitionValues.add(tStringLiteral.value);
+            for (TNullableStringLiteral tStringLiteral : partitionValueList) {
+                if (tStringLiteral.is_null) {
+                    if (partitionType == PartitionType.RANGE) {
+                        throw new AnalysisException("Can't create partition for NULL Range");
```
Comment (zclllyybb) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/PartitionExprUtil.java:134]: add some test

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateTimeV2Literal.java [hunk 3]
```diff
@@ -75,6 +76,10 @@ private void roundMicroSecond(int scale) {
             this.second = localDateTime.getSecond();
             this.microSecond -= 1000000;
         }
+        if (checkRange() || checkDate()) {
+            // may fallback to legacy planner. make sure the behaviour of rounding is same.
```
Comment (zclllyybb) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/literal/DateTimeV2Literal.java:80]: add some test

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/exec/tablet_info.cpp [hunk 4]
```diff
@@ -547,12 +536,13 @@ Status VOlapTablePartitionParam::_create_partition_key(const TExprNode& t_expr,
...
     for (const auto& t_part : partitions) {
-        auto part = _obj_pool.add(new VOlapTablePartition(&_partition_block));
+        auto* part = _obj_pool.add(new VOlapTablePartition(&_partition_block));
         part->id = t_part.id;
         part->is_mutable = t_part.is_mutable;
 
-        DCHECK(t_part.__isset.start_keys == t_part.__isset.end_keys &&
-               t_part.__isset.start_keys != t_part.__isset.in_keys);
+        // we dont pass right keys when it's MAX_VALUE. so may dont have end_keys.
+        //
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:544]: remove the unless code

File: be/src/vec/columns/column_nullable.cpp [hunk 5]
```diff
@@ -390,17 +391,23 @@ int ColumnNullable::compare_at(size_t n, size_t m, const IColumn& rhs_,
     /// instead of the usual nan_direction_hint and is used to implement
     /// the ordering specified by either NULLS FIRST or NULLS LAST in the
     /// ORDER BY construction.
-    const auto& nullable_rhs = assert_cast<const ColumnNullable&>(rhs_);
```
Comment (HappenLee) [inline be/src/vec/columns/column_nullable.cpp:393]: The way change the default behavior in Doris， `compare_at` must use same type column

**Comment by zclllyybb** [comment]
run p0 10

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

File: be/src/exec/tablet_info.cpp [hunk 6]
```diff
@@ -551,16 +504,84 @@ Status VOlapTablePartitionParam::_create_partition_key(const TExprNode& t_expr,
...
+    }
+    part_result->indexes = t_part.indexes;
+    std::sort(part_result->indexes.begin(), part_result->indexes.end(),
+              [](const OlapTableIndexTablets& lhs, const OlapTableIndexTablets& rhs) {
+                  return lhs.index_id < rhs.index_id;
+              });
+    // check index
+    for (int j = 0; j < num_indexes; ++j) {
+        if (part_result->indexes[j].index_id != _schema->indexes()[j]->index_id) {
+            std::stringstream ss;
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:564]: do not use `std::stringstream `

**Review by HappenLee — COMMENTED** [review]
Add doc change

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by BiteTheDDDDt — APPROVED** [review]

File: be/src/vec/common/assert_cast.h [hunk 7]
```diff
@@ -20,25 +20,24 @@
...
-#include "fmt/format.h"
 #include "vec/common/demangle.h"
 
 /** Perform static_cast in release build.
   * Checks type by comparing typeid and throw an exception in debug build.
   * The exact match of the type is checked. That is, cast to the ancestor will be unsuccessful.
   */
 template <typename To, typename From>
-To assert_cast(From&& from) {
+PURE To assert_cast(From&& from) {
```
Comment (zhiqiang-hhhh) [inline be/src/vec/common/assert_cast.h:34]: why add `PURE` here?

**Pull request merged**